tests/system/test_diags.py

```diff
@@ -65,25 +75,49 @@ def setUpClass(self):
                 '/global/u1/f/(.*)/e3sm_diags', '/global/cfs/cdirs/acme/www/{}',
                 TestAllSets.results_dir)
 
-    def check_html_image(self, html_path, png_path):
+    def check_html_image(self, html_path, png_path, full_png_path):
         img_src = None
         option_value = None
         href = None
         with open(html_path, 'r') as html:
             for line in html:
+                # If `img_src` is not defined yet:
                 if not img_src:
                     re_str = '<img src="../../../../{}">'.format(png_path)
                     img_src = re.search(re_str, line)
+                # If `option_value` is not defined yet:
                 if not option_value:
                     re_str = '<option value="../../../../{}">'.format(png_path)
                     option_value = re.search(re_str, line)
+                # If `href` is not defined yet:
                 if not href:
                     re_str = 'href="../../../../{}">'.format(png_path)
                     href = re.search(re_str, line)
         self.assertIsNotNone(img_src)
         self.assertIsNotNone(option_value)
         self.assertIsNotNone(href)
 
+        # https://stackoverflow.com/questions/35176639/compare-images-python-pil
+        path_to_actual_png = full_png_path
+        path_to_expected_png = '../unit_test_images/{}'.format(png_path)
+        actual_png = Image.open(path_to_actual_png).convert('RGB')
+        expected_png = Image.open(path_to_expected_png).convert('RGB')
+        diff = ImageChops.difference(actual_png, expected_png)
+
+        bbox = diff.getbbox()
+        if not bbox:
+            # If `diff.getbbox()` is None, then the images are in theory equal
+            self.assertIsNone(diff.getbbox())
+        else:
+            # Sometimes, a few pixels will differ, but the two images appear identical.
+            # https://codereview.stackexchange.com/questions/55902/fastest-way-to-count-non-zero-pixels-using-python-and-pillow
+            nonzero_pixels = diff.crop(bbox).point(lambda x: 255 if x else 0).convert("L").point(bool).getdata()
+            num_nonzero_pixels = sum(nonzero_pixels)
+            print('\npath_to_actual_png={}'.format(path_to_actual_png))
+            print('path_to_expected_png={}'.format(path_to_expected_png))
+            print('diff has {} nonzero pixels.'.format(num_nonzero_pixels))
+            self.assertTrue(num_nonzero_pixels < 5)
```

Comment:
@chengzhuzhang For some reason, the `lat_lon` actual and expected plots will differ by as many as 4 pixels. My guess is that the plotting software doesn't always draw the contours identically each time. Luckily, a difference this small doesn't appear to actually be visible. Therefore, I believe this check on the number of nonzero pixels is sufficient. Would you agree?

Reply:
Thanks for investigating. Yes, good idea to have this additional check. I think we can proceed with this (<5) threshold and we can refine if needed after trying this on more plots.